src/app.ts

```diff
@@ -236,11 +237,39 @@ export class App {
    *   ctx.res.set('X-Mojo', 'I <3 mojo.js!');
    *   await ctx.render(...args);
    * });
+   *
+   * // Render response with header using nested helper
+   * app.addHelper('renderWith.header', async (ctx, ...args) => {
+   *   ctx.res.set('X-Mojo', 'I <3 mojo.js!');
+   *   await ctx.render(...args);
+   * });
    */
   addHelper(name: string, fn: MojoAction): this {
-    return this.decorateContext(name, function (this: MojoContext, ...args: any[]) {
-      return fn(this, ...args);
-    });
+    const nestedNames = name.split('.');
+    if (nestedNames.length == 1)
+      return this.decorateContext(name, function (this: MojoContext, ...args: any[]) {
+        return fn(this, ...args);
+      });
+    else if (nestedNames.length == 2) {
+      const [getterName, methodName] = nestedNames;
+      this._nestedFunctions[getterName] ??= {};
+      this._nestedFunctions[getterName][methodName] = fn;
+      const ctxScopedFunctions: any = {};
+      let that: unknown;
```

Comment:
I bet this breaks when there is two or more active context instances, each keeping a copy of `const nested = ctx.nested;`, and then calling `nested.fooHelper();` at the same time. Then they would be using each other's stash and you get super hard to debug issues.

Reply:
It actually breaks when you keep a copy of the getter.

For the record, you should be able to reproduce it with this simple Mojo app:

```javascript
/*
 * Minimal helper app to use to check nested helper concurrency
 */
import mojo from '../lib/core.js';

const app = mojo();

app.log.level = 'trace';

app.addHelper('nested.requestId', (ctx) => ctx.req.requestId);

app.get('/:kind', async ctx => {
    const {kind} = ctx.stash;
    const requestId = ctx.req.requestId;
    const nested = ctx.nested; // save nested getter
    const helperOriginalId = kind === 'safe' ? ctx.nested.requestId() : nested.requestId();
    await new Promise(r => setTimeout(r, 1000));
    const helperFinalId = kind === 'safe' ? ctx.nested.requestId() : nested.requestId();
    if (helperOriginalId === requestId && helperFinalId === requestId) {
        ctx.log.trace(`Ok, requestId = ${requestId} everywhere`);
        await ctx.render({ text: `Ok, requestId = ${requestId} everywhere` });
    } else {
        ctx.log.error(
            `Corrupted ctx: requesId: ${requestId} , helper Original: ${helperOriginalId}, helper Final: ${helperFinalId}`);
        await ctx.render({ text: `Not Ok, requesId: ${requestId} , helper Original: ${helperOriginalId}, helper Final: ${helperFinalId}`, status: 500 });
    }
});

app.start();

```
The idea is that a get on `'/safe'` doesn't use the saved getter base, but a get on  `/unsafe` or whatever different from `safe` uses it, and makes it fail.

So this works and logs all Ok traces:
```shell
$> autocannon -c 100 -d 10 -p 10 http://127.0.0.1:3000/safe
```
And this logs a lot of corrupted ctx errors:
```shell
$> autocannon -c 100 -d 10 -p 10 http://127.0.0.1:3000/unsafe
```

I will keep investigating this during the weekend and hopefully push a fix soon

Reply:
In the context class we already use `Proxy` in [two](https://github.com/mojolicious/mojo.js/blob/e95755f6da4b3c6a519c1d4a638130a1816527b4/src/context.ts#L66) [cases](https://github.com/mojolicious/mojo.js/blob/e95755f6da4b3c6a519c1d4a638130a1816527b4/src/context.ts#L198). If you need some inspiration.

Reply:
Thanks, that helped, seems to be the way to go.
I will need some time to refactor several parts. In the meantime I added a test for the concurrency problem, and implemented a naive fix (just builds the functions object inside the getter every time its called)
Again, not squashed yet, to facilitate comparison